README.md
# PdfCover [![Build Status](https://travis-ci.org/xing/pdf_cover.svg?branch=master)](https://travis-ci.org/xing/pdf_cover) [![Code Climate](https://codeclimate.com/github/xing/pdf_cover/badges/gpa.svg)](https://codeclimate.com/github/xing/pdf_cover) [![Coverage Status](https://coveralls.io/repos/github/xing/pdf_cover/badge.svg?branch=master)](https://coveralls.io/github/xing/pdf_cover?branch=master)

With this gem you can easily have attachments for PDF files that have associated
images generated for their first page.

Support is provided both for [Paperclip](https://github.com/thoughtbot/paperclip)
and [CarrierWave](https://github.com/carrierwaveuploader/carrierwave).

In both cases the JPEG quality and resolution are optional and will be set to 85%
and 300dpi respectively when not provided.

## Paperclip Support

To add a PDF cover style to your attachments you can do something like this:

```Ruby
class WithPaperclip < ActiveRecord::Base
  include PdfCover

  pdf_cover_attachment :pdf, styles: { pdf_cover: ['', :jpeg]},
    convert_options: { all: '-quality 95 -density 300' },

  validates_attachment_content_type :pdf, content_type: %w(application/pdf)
end
```

This will define an attachment called `pdf` which has a `pdf_cover` style attached
to it that is a JPEG of the first page in the PDF. You can pass any option that you
would normally pass to `has_attached_file` in the options hash and it will be
passed through to the underlying `has_attached_file` call.

## CarrierWave

When using CarrierWave you can implement this gem's functionality
you can do something like this:

```Ruby
class WithCarrierwaveUploader < CarrierWave::Uploader::Base
  include PdfCover

  storage :file

  version :image do
    pdf_cover_attachment quality: 95, resolution: 300
  end
end
```

In this case, when we mix the `PdfCover` module in, it adds the `pdf_cover_attachment`
method to our uploader. We only need to call it inside one of our versions to get the
pdf to image feature.

# Developing this gem

After cloning this gem locally just run the `bin/setup` script to set everything
up. This will:

- Run `bundle` to install the development dependencies.
- Initialize the database used by the `spec/dummy` rails application that
we use to test the ActiveRecord+(Paperclip|CarrierWave) integration.
- Generate JPEG files used in the specs.

## Running the specs

Once you have setup the gem locally you can just run `rake` from the root folder
of the gem to run the specs.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2021-02-11
### Breaking change
- Drop support for Rails 4.x [Pull Request](https://github.com/xing/pdf_cover/pull/12)

## [0.3.2] - 2020-11-06
### Changed
- Update dependencies with security vulnerabilities detected [Pull Request](https://github.com/xing/pdf_cover/pull/11)

## [0.3.1] - 2019-11-26
### Added
- Enforce JPEG content type for carrierwave [Pull Request](https://github.com/xing/pdf_cover/pull/10)

## [0.2.4] - 2017-10-18
### Added
- Handle quotes in windows when calling GS [Pull Request](https://github.com/xing/pdf_cover/pull/9)

## [0.2.3] - 2017-08-21
### Added
- Remove `dTextAlphaBits` parameter to GS [Pull Request](https://github.com/xing/pdf_cover/pull/8)

## [0.2.3] - 2017-08-21
### Added
- Remove `dTextAlphaBits` parameter to GS [Pull Request](https://github.com/xing/pdf_cover/pull/8)

## [0.2.2] - 2017-07-05
### Added
- Add support for rails 5 [Pull Request](https://github.com/xing/pdf_cover/pull/7)

## [0.2.1] - 2016-10-24
### Added
- Add default antialias [Pull Request](https://github.com/xing/pdf_cover/pull/6)

## [0.2.0] - 2016-06-13
### Added
- Raise new error if GS execution fails [Pull Request](https://github.com/xing/pdf_cover/pull/5)

## [0.1.4] - 2016-05-12
### Fixed
- Fix issue with uppercase file names in Carrierwave [Pull Request](https://github.com/xing/pdf_cover/pull/4)

## [0.1.3] - 2016-03-02
### Fixed
- Fix naming conflict between PdfCovers
- Fix namespace issue [Pull Request](https://github.com/xing/pdf_cover/pull/3)

## [0.1.2] - 2016-03-01
### Fixed
- Remove nil parameters on CarrierWave inclusion

## [0.1.1] - 2016-03-01
### Added
- Allow to set quality and resolution of result image
- New default quality and resolution for jpegs
- Require rmagick on a single place

## [0.1.0] - 2016-02-25
Initial release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)

ENV["RAILS_ENV"] ||= "test"

# Code coverage
require "coveralls"
Coveralls.wear!

# Development dependencies
require "byebug"
require "paperclip"
require "paperclip/matchers"
require "carrierwave"
require "RMagick"

# The gem itself
require "pdf_cover"

RSpec.configure do |config|
  config.include Paperclip::Shoulda::Matchers

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  config.disable_monkey_patching!

  config.profile_examples = 10
  config.order = :random
  config.default_formatter = "Fivemat"
  config.expose_dsl_globally = true
end


spec/generate_jpegs.sh
#!/bin/bash

# NOTE: Update the converter.rb file when changing these values
DEFAULT_FORMAT="jpeg"
DEFAULT_QUALITY=85
DEFAULT_RESOLUTION=300
DEFAULT_ANTIALIASING=4

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/test_pdfs"

echo "Working directory $DIR"

echo "Removing previous output"
rm $DIR/*.jpg

IFS="
"

for INPUT_FILE in $(cd $DIR && ls *.pdf)
do
  OUTPUT_FILE="$DIR/${INPUT_FILE/.pdf/.jpg}"
  echo "Generating $OUTPUT_FILE from $INPUT_FILE"
  gs -sOutputFile="$OUTPUT_FILE" \
    -dNOPAUSE \
    -sDEVICE="$DEFAULT_FORMAT" \
    -dJPEGQ=$DEFAULT_QUALITY \
    -dFirstPage=1 \
    -dLastPage=1 \
    -r$DEFAULT_RESOLUTION \
    -dGraphicsAlphaBits=$DEFAULT_ANTIALIASING \
    -q "$DIR/$INPUT_FILE" -c quit
done


spec/dummy/db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20160208154250) do

  create_table "with_carrierwaves", force: :cascade do |t|
    t.string "pdf"
  end

  create_table "with_paperclips", force: :cascade do |t|
    t.string   "pdf_file_name"
    t.string   "pdf_content_type"
    t.integer  "pdf_file_size"
    t.datetime "pdf_updated_at"
  end

end


lib/pdf_cover.rb
require "pdf_cover/version"
require "pdf_cover/converter"

# This module provides methods for CarrierWave::Uploader::Base subclasses and
# for ActiveRecord models that want to include attachments to simplify the
# generation of JPEG images from the first page of a PDF file that is uploaded
# by the users.
# Include this module in your class and check the `ClassMethods` documentation
# that corresponds to your attachments managing library in this same file.
module PdfCover
  module ClassMethods
    module CarrierWave
      # When called in the context of a CarrierWave::Uploader::Base subclass,
      # this method will add a processor to the current attachment or version
      # that generates a JPEG with 95 quality from the first page of the given
      # PDF.
      #
      # This will not make any validation on the given content type, you must
      # [do it yourself](https://github.com/carrierwaveuploader/carrierwave#securing-uploads)
      # on your uploader.
      #
      # @example Adding a version to your uploader:
      #   class WithCarrierwaveUploader < CarrierWave::Uploader::Base
      #     include PdfCover
      #
      #     storage :file
      #
      #     version :image do
      #       pdf_cover_attachment
      #     end
      #   end
      def pdf_cover_attachment(options = {})
        process pdf_cover: [options[:quality], options[:resolution]]
        process enforce_content_type: "image/jpeg"

        define_method :full_filename do |for_file = model.logo.file|
          for_file.gsub(/pdf$/i, "jpeg")
        end
      end
    end

    module Paperclip
      # Adds a new attached file to the caller that has the pdf cover processors
      # prepended to the list of processors given.
      #
      # @param attachment_name [Symbol] the name of the new attachment. The fields
      # described in the Paperclip documentation for attachments are needed for this one.
      # @param options [Hash] the same options that are accepted by Paperclip, they
      # are passed to the has_attached_file call with just a new processor
      # prepended to the given ones. The PdfCover processor will use the quality
      # provided in the `convert_options` option when generating the jpeg.
      #
      # @example A sample ActiveRecord model with a pdf_cover attachment:
      #   class WithPaperclip < ActiveRecord::Base
      #     include PdfCover
      #
      #     pdf_cover_attachment :pdf, styles: { pdf_cover: ['', :jpeg]},
      #       convert_options: { all: '-quality 95 -density 300' }
      #
      #     # Note that you must set content type validation as required by Paperclip
      #     validates_attachment_content_type :pdf, content_type: %w(application/pdf)
      #   end
      def pdf_cover_attachment(attachment_name, options = {})
        options[:processors] = (options[:processors] || []).unshift(:pdf_cover_processor)

        has_attached_file attachment_name, options
      end
    end
  end

  class << self
    def included(base)
      if carrierwave_defined?(base)
        base.extend ClassMethods::CarrierWave
      elsif paperclip_defined?
        require "paperclip/pdf_cover_processor"
        base.extend ClassMethods::Paperclip
      else
        fail "#{base} is not a CarrierWave::Uploader and Paperclip is not defined ¯\\_(ツ)_/¯"
      end
    end

    private

    def carrierwave_defined?(base)
      defined?(CarrierWave::Uploader::Base) && base.ancestors.include?(CarrierWave::Uploader::Base)
    end

    def paperclip_defined?
      defined?(Paperclip)
    end
  end

  # This is the method used by the CarrierWave processor mechanism
  def pdf_cover(quality, resolution)
    options = { quality: quality, resolution: resolution }.compact
    converted_file = PdfCover::Converter.new(file, options).converted_file
    FileUtils.cp(converted_file, current_path)
  end

  def enforce_content_type(content_type)
    file.content_type = content_type
  end
end


lib/paperclip/pdf_cover_processor.rb
if Kernel.const_defined?(:Paperclip)
  require "pdf_cover/converter"

  module Paperclip
    # This is a Paperclip::Processor that can be used to generate an image
    # from a PDF file. The image is only of the first page.
    #
    # We inherit the following instance variables from Paperclip::Processor:
    #   @file the file that will be operated on (which is an instance of File)
    #   @options a hash of options that were defined in has_attached_file's style hash
    #   @attachment the Paperclip::Attachment itself
    class PdfCoverProcessor < Processor
      QUALITY_CONVERT_OPTION_REGEX = /-quality\s+(?<quality>\d+)/
      RESOLUTION_CONVERT_OPTION_REGEX = /-density\s+(?<resolution>\d+)/

      def make
        ::PdfCover::Converter.new(@file, converter_options).converted_file
      end

      def converter_options
        format.merge(jpeg_quality).merge(jpeg_resolution)
      end

      def format
        { format: @options[:format].to_s }
      end

      def jpeg_quality
        extract_convert_option(:quality, QUALITY_CONVERT_OPTION_REGEX)
      end

      def jpeg_resolution
        extract_convert_option(:resolution, RESOLUTION_CONVERT_OPTION_REGEX)
      end

      def extract_convert_option(key, regex)
        match_data = regex.match(@options[:convert_options])
        match = match_data && match_data[key]

        match ? { key => match } : {}
      end
    end
  end
end


lib/pdf_cover/converter.rb
require "open3"

module PdfCover
  class Converter
    # NOTE: Update the generate_jpegs.sh script when changing these values
    DEFAULT_FORMAT = "jpeg".freeze
    DEFAULT_QUALITY = 85
    DEFAULT_RESOLUTION = 300
    DEFAULT_ANTIALIASING = 4

    COMMAND_EXECUTION_SUCCESS_CODE = 0
    COMMAND_NOT_FOUND_CODE = 127 # @see http://www.tldp.org/LDP/abs/html/exitcodes.html

    class CommandFailedError < StandardError
      def initialize(stdout_str, stderr_str)
        super("PDF conversion failed:\nSTDOUT: #{stdout_str}\nSTDERR: #{stderr_str}")
      end
    end

    class CommandNotFoundError < StandardError
      def initialize
        super("Could not run the `gs` command. Make sure that GhostScript is in your PATH.")
      end
    end

    class InvalidOption < StandardError
      def initialize(option_name, option_value)
        super("Invalid option '#{option_name}' with value: '#{option_value}'")
      end
    end

    def initialize(file, options = {})
      @file = file
      extract_options(options)
    end

    # @raises PdfCover::Converter::CommandNotFoundError if GhostScript is not found
    # @raises PdfCover::Converter::CommandFailedError if GhostScript returns a non-zero status
    def converted_file
      parameters = build_parameters(file_path, device)
      stdout_str, stderr_str, status = execute_command("gs #{parameters}")

      case status
        when COMMAND_EXECUTION_SUCCESS_CODE then destination_file
        when COMMAND_NOT_FOUND_CODE then fail CommandNotFoundError
        else fail CommandFailedError.new(stdout_str, stderr_str)
      end
    rescue => e
      destination_file.close
      raise e
    end

    private

    def basename
      @basename ||= File.basename(@file.path, File.extname(@file.path))
    end

    def build_parameters(source, device)
      %W(-sOutputFile="#{destination_path}" -dNOPAUSE
         -sDEVICE="#{device}" -dJPEGQ=#{@quality}
         -dFirstPage=1 -dLastPage=1
         -dGraphicsAlphaBits=#{@antialiasing}
         -r#{@resolution} -q "#{source}"
         -c quit)
        .join(" ")
    end

    def destination_file
      @destination_file ||= Tempfile.new([basename, ".#{@format}"]).tap do |file|
        file.binmode
      end
    end

    def destination_path
      File.expand_path(destination_file.path)
    end

    def device
      @format.to_s == "jpg" ? "jpeg" : @format.to_s
    end

    def execute_command(command)
      Open3.capture3(command)
    end

    def extract_options(options)
      @format = options.fetch(:format, DEFAULT_FORMAT)
      @quality = options.fetch(:quality, DEFAULT_QUALITY).to_i
      @resolution = options.fetch(:resolution, DEFAULT_RESOLUTION).to_i
      @antialiasing = options.fetch(:antialiasing, DEFAULT_ANTIALIASING).to_i

      validate_options
    end

    def file_path
      @file_path ||= File.expand_path(@file.path)
    end

    def validate_options
      fail InvalidOption.new(:format, @format) unless %w(jpg jpeg png).include?(@format)
      fail InvalidOption.new(:quality, @quality) unless @quality.between?(1, 100)
      fail InvalidOption.new(:resolution, @resolution) unless @resolution > 1
      fail InvalidOption.new(:antialiasing, @antialiasing) unless @antialiasing >= 1 && @antialiasing <= 4
    end
  end
end
